File: fusion/exposure_fusion/src/exposure_fusion.cpp
// C++
#include <map>
#include <vector>
#include <chrono>
#include <iostream>
#include <functional>
// OpenCV
#include <opencv2/opencv.hpp>


namespace {
    void run(const std::function<void()>& work=[]{}, const std::string message="") {
        auto start = std::chrono::steady_clock::now();
        work();
        auto finish = std::chrono::steady_clock::now();
        auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(finish - start);
        std::cout << message << " " << duration.count() << " ms" <<  std::endl;
    }

    void cv_show(const cv::Mat& one_image, const char* info="") {
        cv::imshow(info, one_image);
        cv::waitKey(0);
        cv::destroyAllWindows();
    }

    bool cv_write(const cv::Mat& source, const std::string save_path) {
        return cv::imwrite(save_path, source, std::vector<int>({cv::IMWRITE_PNG_COMPRESSION, 0}));
    }

    void cv_info(const cv::Mat& one_image) {
        std::cout << "高  :  " << one_image.rows << "\n宽  :  " << one_image.cols << "\n通道 :  " << one_image.channels() << std::endl;
        std::cout << "步长 :  " << one_image.step << std::endl;
        std::cout << "是否连续" << std::boolalpha << one_image.isContinuous() << std::endl;
    }

    cv::Mat cv_concat(const std::vector<cv::Mat> images, const bool v=false) {
        cv::Mat result;
        if(not v) cv::hconcat(images, result);
        else cv::vconcat(images, result);
        return result;
    }

    cv::Mat make_pad(const cv::Mat& one_image, const int pad_H, const int pad_W) {
        cv::Mat padded_image;
        cv::copyMakeBorder(one_image, padded_image, pad_H, pad_H, pad_W, pad_W, cv::BORDER_REPLICATE);
        return padded_image;
    }

    template<typename T>
    cv::Mat toint8(const std::vector<T>& source, const int H, const int W, const int C, const int _type, const double times=2) {
        cv::Mat result(H, W, _type);
        const int length = H * W * C;
        for(int i = 0;i < length; ++i) result.data[i] = cv::saturate_cast<uchar>(std::abs(source[i]) * times);
        return result;
    }
}



namespace {
    inline double fast_exp(const double y) {
        double d;
        *(reinterpret_cast<int*>(&d) + 0) = 0;
        *(reinterpret_cast<int*>(&d) + 1) = static_cast<int>(1512775 * y + 1072632447);
        return d;
    }

    inline double square(const double x) {
        return x * x;
    }
}


std::map<const std::string, cv::Mat> exposure_fusion(
        const std::vector<cv::Mat>& sequence,
        const std::tuple<float, float, float>& alphas={1.0, 1.0, 1.0},
        const bool use_lappyr=true,
        const int layers_num=5,
        const float best_illumination=0.5,
        const double sigma=0.2) {
    // 准备一些中间计算所需变量
    const double sigma_inv = 1. / (2 * sigma * sigma);
    const float norm = 1.f / 255;
    // 获取图像信息
    const int H = sequence.front().rows;
    const int W = sequence.front().cols;
    const int C = sequence.front().channels();
    assert(C == 3 and "该算法只支持 RGB24 图像 !");
    for(const auto& image : sequence)
        assert(image.rows == H and image.cols == W and image.channels() == C and "图像序列的大小必须都一致!");
    const int length = H * W;
    // 看有几张图像
    const int sequence_len = sequence.size();
    // 根据先验知识, 计算对比度、饱和度、亮度, 求权重
    std::vector< cv::Mat > weights;
    for(const auto& image : sequence) {
        // 【1】根据对比度求权重
        cv::Mat gray;
        cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
        const auto gray_padded = make_pad(gray, 1, 1);
        std::vector<float> contrast(length, 0);
        const int W_2 = W + 2;
        for(int i = 1; i < H + 1; ++i) {
            const uchar* const row_ptr = gray_padded.data + i * W_2;
            float* const res_ptr = contrast.data() + (i - 1) * W;
            for(int j = 1; j < W + 1; ++j)
                res_ptr[j - 1] = norm * std::abs(row_ptr[j - 1] + row_ptr[j + 1] + row_ptr[j - W_2] + row_ptr[j + W_2] - 4 * row_ptr[j]);
        }
        // 【2】根据饱和度求权重
        std::vector<float> saturation(length, 0);
        for(int i = 0; i < H; ++i) {
            const uchar* const row_ptr = image.data + i * W * 3;
            float* const res_ptr = saturation.data() + i * W;
            for(int j = 0; j < W; ++j) {
                const int p = 3 * j;
                const float mean = (row_ptr[p] + row_ptr[p + 1] + row_ptr[p + 2]) * norm / 3;
                res_ptr[j] = std::sqrt((square(row_ptr[p] * norm - mean) + square(row_ptr[p + 1] * norm - mean) + square(row_ptr[p + 2] * norm - mean)) * 1.f / 3);
            }
        }
        // 【3】根据亮度求权重
        std::vector<float> illumination(length, 0);
        for(int i = 0; i < H; ++i) {
            const uchar* const row_ptr = image.data + i * W * 3;
            float* const res_ptr = illumination.data() + i * W;
            for(int j = 0; j < W; ++j) {
                const int p = 3 * j;
                res_ptr[j] = fast_exp(-square(row_ptr[p] * norm - best_illumination) * sigma_inv)
                           * fast_exp(-square(row_ptr[p + 1] * norm - best_illumination) * sigma_inv)
                           * fast_exp(-square(row_ptr[p + 2] * norm - best_illumination) * sigma_inv);
            }
        }
        /*
        cv_show(image);
        cv_show(cv_concat({
            toint8(contrast, H, W, 1, CV_8UC1, 255.0),
            toint8(saturation, H, W, 1, CV_8UC1, 255.0),
            toint8(illumination, H, W, 1, CV_8UC1, 255.0)
        }));
        */
        // 三者结合, 求权重
        cv::Mat cur_weight(H, W, CV_32F);
        float* const weight_ptr = cur_weight.ptr<float>();
        for(int i = 0; i < length; ++i)
            weight_ptr[i] = std::pow(contrast[i], std::get<0>(alphas)) *
                            std::pow(saturation[i], std::get<1>(alphas)) *
                            std::pow(illumination[i], std::get<2>(alphas));
        weights.emplace_back(cur_weight);
    }
    std::vector<float*> weights_ptrs;
    for(int k = 0;k < sequence_len; ++k) weights_ptrs.emplace_back(weights[k].ptr<float>());
    // 求归一化的权重
    for(int i = 0;i < length; ++i) {
        float weight_sum = 1e-12;
        for(int k = 0;k < sequence_len; ++k) weight_sum += weights_ptrs[k][i];
        for(int k = 0;k < sequence_len; ++k) weights_ptrs[k][i] /= weight_sum;
    }
    // for(int k = 0;k < sequence_len; ++k) cv_show(toint8(weights[k], H, W, 1, CV_8UC1, 255));
    // 准备融合结果
    std::map<const std::string, cv::Mat> results;
    // 粗糙的融合
    auto weighted_fusion = [&](const bool use_gaussi=false)
            ->cv::Mat {
        cv::Mat fused = cv::Mat::zeros(H, W, sequence.front().type());
        uchar* fused_ptr = fused.data;
        // 每张图像对应一个权重图
        for(int k = 0;k < sequence_len; ++k) {
            // 对权重图做高斯模糊
            cv::Mat blurred;
            if(use_gaussi) cv::GaussianBlur(weights[k], blurred, cv::Size(49, 49), 8, 8);
            else blurred = std::ref(weights[k]);
            const float* const w_ptr = blurred.ptr<float>();
            const uchar* const cur_image = sequence[k].data;
            for(int i = 0;i < length; ++i) {
                const int p = 3 * i;
                fused_ptr[p] += w_ptr[i] * cur_image[p];
                fused_ptr[p + 1] += w_ptr[i] * cur_image[p + 1];
                fused_ptr[p + 2] += w_ptr[i] * cur_image[p + 2];
            }
        }
        const int total_length = 3 * length;
        for(int i = 0;i < total_length; ++i) fused_ptr[i] = cv::saturate_cast<uchar>(fused_ptr[i]);
        return fused;
    };
    results.emplace("naive", weighted_fusion());
    results.emplace("gaussi_smoothed", weighted_fusion(true));
    // 是否使用拉普拉斯金字塔融合
    if(use_lappyr) {
        // 检查 layers_num 不要太离谱
        assert((1 << layers_num) < sequence[0].cols and (1 << layers_num) < sequence[0].rows and "layers 太大, 图像分辨率太低不支持");
        // 从 high_res 开始构建层数为 layers_num 的高斯金字塔
        auto build_gaussi_pyramids = [](const cv::Mat& high_res, const int layers_num)
                -> std::vector<cv::Mat> {
            std::vector<cv::Mat> this_flash({high_res});
            for(int i = 1; i < layers_num; ++i) {
                cv::Mat blurred;
                cv::GaussianBlur(this_flash[i - 1], blurred, cv::Size(5, 5), 0.83, 0.83);
                cv::resize(blurred, blurred, cv::Size(this_flash[i - 1].cols / 2, this_flash[i - 1].rows / 2));
                this_flash.emplace_back(blurred);
            }
            return this_flash;
        };
        // 求每张图的权重的高斯金字塔
        std::vector< std::vector<cv::Mat> > sequence_weights_pyramids;
        sequence_weights_pyramids.reserve(sequence_len);
        cv::Mat high_res(H, W, CV_32FC1);
        for(int k = 0;k < sequence_len; ++k)
            sequence_weights_pyramids.emplace_back(build_gaussi_pyramids(weights[k], layers_num));
        // 提前释放 weights 空间
        for(int k = 0;k < sequence_len; ++k) weights[k].release();
        // 求每张图的高斯金字塔
        std::vector< std::vector<cv::Mat> > sequence_gaussi_pyramids;
        sequence_gaussi_pyramids.reserve(sequence_len);
        for(int k = 0;k < sequence_len; ++k)
            sequence_gaussi_pyramids.emplace_back(build_gaussi_pyramids(sequence[k], layers_num));
        // 转换数据类型 uchar -> float, 因为后面有减法
        for(int k = 0;k < sequence_len; ++k)
            for(int i = 0;i < layers_num; ++i)
                sequence_gaussi_pyramids[k][i].convertTo(sequence_gaussi_pyramids[k][i], CV_32F);
        // 求每张图的 laplace 金字塔
        std::vector< std::vector<cv::Mat> > sequence_laplace_pyramids;
        sequence_laplace_pyramids.reserve(sequence_len);
        for(int k = 0;k < sequence_len; ++k) {
            auto& gaussi_pyramid = sequence_gaussi_pyramids[k];
            cv::Mat upsampled = gaussi_pyramid[layers_num - 1].clone();
            std::vector<cv::Mat> pyramid({upsampled});
            for(int i = layers_num - 1; i > 0; --i) {
                cv::resize(gaussi_pyramid[i], upsampled, cv::Size(gaussi_pyramid[i - 1].cols, gaussi_pyramid[i - 1].rows));
                pyramid.emplace_back(gaussi_pyramid[i - 1] - upsampled);
            }
            std::reverse(pyramid.begin(), pyramid.end());
            sequence_laplace_pyramids.emplace_back(pyramid);
            for(int i = 0;i < layers_num; ++i) gaussi_pyramid[i].release(); // 释放 gaussi_pyramid 空间, 反正没用了
        }
        // 每一个尺度, 融合一系列图像的的 laplace 细节, 得到一个融合的 laplace 金字塔
        std::vector<cv::Mat> fused_laplace_pyramid;
        fused_laplace_pyramid.reserve(layers_num);
        for(int i = 0;i < layers_num; ++i) {
            cv::Mat weighted_laplace = cv::Mat::zeros(sequence_weights_pyramids[0][i].rows, sequence_weights_pyramids[0][i].cols, CV_32FC3);
            for(int k = 0; k < sequence_len; ++k) {
                cv::Mat new_weights;
                std::vector<cv::Mat> new_weights_vector({sequence_weights_pyramids[k][i], sequence_weights_pyramids[k][i], sequence_weights_pyramids[k][i]});
                cv::merge(new_weights_vector, new_weights);
                weighted_laplace += new_weights.mul(sequence_laplace_pyramids[k][i]); // 这里慢了, 拷贝的消耗大, 直接从指针赋值可能快一点
            }
            fused_laplace_pyramid.emplace_back(weighted_laplace);
        }
        // 从最底层开始, 每次上采样加上同等尺度的 laplace 细节
        cv::Mat fused = fused_laplace_pyramid[layers_num - 1];
        for(int i = layers_num - 2; i >= 0; --i) {
            cv::Mat upsampled;
            cv::resize(fused, upsampled, cv::Size(fused_laplace_pyramid[i].cols, fused_laplace_pyramid[i].rows));
            fused = upsampled + fused_laplace_pyramid[i];
        }
        fused.convertTo(fused, CV_8U);
        results.emplace("laplace_pyramid", fused);
    }
    return results;
}


 #include <ghc/filesystem.hpp>


int main() {

    // 获取图像列表
    std::vector<cv::Mat> sequence;
    const std::string sequence_dir("./images/input/5/");
    auto sequence_list = ghc::filesystem::directory_iterator(sequence_dir);
    for(const auto& it : sequence_list) {
        cv::Mat current = cv::imread(sequence_dir + it.path().filename().string());
        if(current.empty()) continue;
        sequence.emplace_back(current);
    }
    // 如果 layers_num 降低为 5, 就会有很严重的光晕
    const auto fusion_result = exposure_fusion(sequence, {1.0, 1.0, 1.0}, true, 7);
    for(const auto &item : fusion_result) {
        cv_show(item.second, item.first.c_str());
        cv_write(item.second, std::string("./images/output/5/C++_" + item.first + ".png"));
    }
    return 0;
}

File: fusion/exposure_fusion/src/exposure_fusion.py
# Python
import os
import sys
import math
import copy
# 3rd party
import cv2
import numpy


def visualize(sequence_pyramid):
	S2 = len(sequence_pyramid)
	layers_num_2 = len(sequence_pyramid[0])
	_size = sequence_pyramid[0][0].shape[1::-1]
	display = numpy.concatenate(
		[cv2.normalize(numpy.concatenate([cv2.resize(sequence_pyramid[s][i], _size) for i in range(layers_num_2)], axis=0), None, alpha=0, beta=1, norm_type=cv2.NORM_MINMAX) for s in range(S2)], axis=1)
	global save_dir
	cv2.imwrite(os.path.join(save_dir, "sequence_laplace_pyramids.png"), (display * 255).astype("uint8"), [cv2.IMWRITE_PNG_COMPRESSION, 0])


def cv_show(image, message="crane"):
	cv2.imshow(message, image)
	cv2.waitKey(0)
	cv2.destroyAllWindows()



def exposure_fusion(sequence, alphas=(1.0, 1.0, 1.0), best_illumination=0.5, sigma=0.2, eps=1e-12, 
		use_lappyr=True, use_gaussi=False, layers_num=5, scale=2.0, visual=False):
	# 转化成 float 数据
	sequence = numpy.stack([it.astype("float32") / 255 for it in sequence], axis=0)
	S = len(sequence)
	H, W, C = sequence[0].shape
	# 准备一些中间变量
	laplace_kernel = numpy.array(([0,  1, 0], [1, -4, 1], [0,  1, 0]), dtype="float32")
	mse = lambda l, r: (l - r) * (l - r)
	best_illumination = numpy.full((H, W), best_illumination, dtype='float32')
	# 存放每张图像的权重图
	weights = []
	# 每张图像都求一个初始权重
	for s in range(S):
		# 从拉普拉斯求对比度
		gray = cv2.cvtColor(sequence[s], cv2.COLOR_BGR2GRAY)
		contrast = cv2.filter2D(gray, -1, laplace_kernel, borderType=cv2.BORDER_REPLICATE)
		contrast = numpy.abs(contrast)
		# 求饱和度
		mean = numpy.mean(sequence[s], axis=-1)
		saturation = numpy.sqrt(numpy.mean([mse(sequence[s][:, :, ch], mean) for ch in range(3)], axis=0))
		# 求亮度
		illumination = [numpy.exp(-0.5 * mse(sequence[s][:, :, ch], best_illumination) / (sigma * sigma)) for ch in range(3)]
		illumination = numpy.prod(illumination, axis=0)
		# 三者加权
		cur_weight = numpy.power(contrast, alphas[0]) * numpy.power(saturation, alphas[1]) * numpy.power(illumination, alphas[2])
		weights.append(cur_weight)
	# 归一化
	normalize = lambda x: x / numpy.expand_dims(numpy.sum(x, axis=0), axis=0)
	weights = numpy.stack(weights, axis=0)
	weights += eps
	weights = normalize(weights)
	# 这里要把 sequence 还原回来
	sequence *= 255;
	# 粗糙的融合
	origin_fusion = numpy.sum(numpy.stack([weights, weights, weights], axis=-1) * sequence, axis=0)
	origin_fusion = numpy.clip(origin_fusion, 0, 255).astype("uint8")
	results = {"naive": origin_fusion}
	# 是否用高斯
	if(use_gaussi == True):
		# 使用高斯模糊对 weights 做模糊
		smoothed_weights = numpy.stack([cv2.GaussianBlur(w, (49, 49), 8) for w in weights], axis=0)
		smoothed_weights = normalize(smoothed_weights)
		smoothed_fusion = numpy.sum(numpy.stack([smoothed_weights, smoothed_weights, smoothed_weights], axis=-1) * sequence, axis=0)
		smoothed_fusion = numpy.clip(smoothed_fusion, 0, 255).astype("uint8")
		results.update({"gaussi_smoothed": smoothed_fusion})
	# 是否用拉普拉斯
	if(use_lappyr == True):

		# 根据最高分辨率的图像 high_res, 得到高度 layers 的高斯金字塔
		def build_gaussi_pyramid(high_res, layers):
			this_flash = [high_res]
			for i in range(1, layers):
				# 先对当前权重做高斯模糊, 然后下采样
				blurred = cv2.GaussianBlur(this_flash[i - 1], (5, 5), 0.83)
				blurred = blurred[::2, ::2]
				this_flash.append(blurred)
			return this_flash

		# 根据已知的高斯金字塔, 从最底层开始上采样, 得到每一个尺度的 laplace 细节
		def build_laplace_pyramaid(gaussi_pyramid, layers):
			upsampled = gaussi_pyramid[layers - 1]
			pyramid = [upsampled]
			for i in range(layers - 1, 0, -1):
				size = (gaussi_pyramid[i - 1].shape[1], gaussi_pyramid[i - 1].shape[0])
				# upsampled = cv2.resize(upsampled, size) # 假如我一直上采样呢
				upsampled = cv2.resize(gaussi_pyramid[i], size)
				pyramid.append(gaussi_pyramid[i - 1] - upsampled)
			pyramid.reverse() # 目前分辨率都是从高到低排列的
			return pyramid

		# 求每张图的权重的高斯金字塔
		sequence_weights_pyramids = [build_gaussi_pyramid(weights[s], layers_num) for s in range(S)]
		# 求每张图的高斯金字塔, 以求 laplace
		sequence_gaussi_pyramids = [build_gaussi_pyramid(sequence[s], layers_num) for s in range(S)]
		# 求每张图的 laplace 金字塔
		sequence_laplace_pyramids = [build_laplace_pyramaid(sequence_gaussi_pyramids[s], layers_num) for s in range(S)]
		# 这里可以归一化, 做可视化
		if(visual): visualize(sequence_laplace_pyramids)
		# 每一个尺度, 融合一系列图像的的 laplace 细节, 得到一个融合的 laplace 金字塔
		fused_laplace_pyramid = [numpy.sum([sequence_laplace_pyramids[k][n] * 
				numpy.atleast_3d(sequence_weights_pyramids[k][n]) for k in range(S)], axis=0) for n in range(layers_num)]

		# 先从最底层的图像开始, 每次上采样都加上同等尺度的 laplace 细节
		start = fused_laplace_pyramid[layers_num - 1]
		for i in range(layers_num - 2, -1, -1):
			upsampled = cv2.resize(start, (fused_laplace_pyramid[i].shape[1], fused_laplace_pyramid[i].shape[0]))
			start = fused_laplace_pyramid[i] + upsampled
		# 灰度值截断在 0-255 之间
		start = numpy.clip(start, 0, 255).astype("uint8")
		# 放到结果列表中
		results.update({"laplace_pyramid": start})

	return results



images_dir = "../images/input/5"
save_dir = "../images/output/5"
os.makedirs(save_dir, exist_ok=True)

# 读取图片
images_list = [os.path.join(images_dir, it) for it in os.listdir(images_dir)]
sequence = numpy.stack([cv2.imread(name) for name in images_list])

# 曝光融合
fused_results = exposure_fusion(
	sequence, 
	use_lappyr=True, 
	use_gaussi=True, 
	layers_num=7,
	alphas=(1.0, 1.0, 1.0),
	visual=False)

# 展示与保存
for l, r in fused_results.items():
	cv_show(r, l)
	cv2.imwrite(os.path.join(save_dir, l + ".png"), r, [cv2.IMWRITE_PNG_COMPRESSION, 0])